File: Tools/flashtool/MainWindow.py
from Connection import connection
from Dispatcher import dispatcher

from copy import copy
from datetime import datetime

from PyQt4 import uic
from PyQt4.QtCore import *
from PyQt4.QtGui import *
from DockWindows import *

class MainWindow(QMainWindow):
    # TODO:
    # * Layout Load/Save with recently used list
    # * Lock Layout
    #   * https://github.com/tegesoft/qt-manhattan-style ?
    #

    def __init__(self):
        QMainWindow.__init__(self)
        uic.loadUi("MainWindow.ui", self)

        self.setDockNestingEnabled(True)

        self.statisticsLabel = QLabel()
        self.connectionLabel = QLabel()
        self.statusBar.addPermanentWidget(self.statisticsLabel)
        self.statusBar.addWidget(self.connectionLabel)

        self.statusBar.setStyleSheet("QStatusBar::item { border: none; }")
        self.centralwidget.setStyleSheet("background-image: url(Images/carbon_gradient.png)")

        self.action_full_screen.triggered.connect(self.action_full_screen_triggered)

        self.add_dock_window(ConnectionWindow(), Qt.LeftDockWidgetArea)
        self.add_dock_window(UpdateWindow(), Qt.RightDockWidgetArea)
        self.add_dock_window(ConsoleWindow(), Qt.BottomDockWidgetArea)
        self.add_dock_window(GLWindow(), Qt.BottomDockWidgetArea)

        settings = QSettings()
        self.restoreGeometry(settings.value("windowGeometry").toByteArray())
        self.restoreState(settings.value("windowState").toByteArray())

        self.action_full_screen.setChecked(self.windowState() & Qt.WindowFullScreen)

        self.old_stats = None
        self.timer = QTimer()
        self.timer.timeout.connect(self.update_gui)
        self.timer.start(500)

    def closeEvent(self, event):
        settings = QSettings()
        settings.setValue("windowGeometry", self.saveGeometry())
        settings.setValue("windowState", self.saveState())
        event.accept()

    def add_dock_window(self, window, area):
        window.setWindowFlags(Qt.Widget)

        dock = QDockWidget()
        dock.setObjectName(window.windowTitle())
        dock.setWindowTitle(window.windowTitle())
        dock.setWidget(window)

        self.addDockWidget(area, dock)

        self.menu_view.insertAction(
            self.menu_view.actions()[-2],
            dock.toggleViewAction()
        )

    def action_full_screen_triggered(self):
        if self.action_full_screen.isChecked():
            self.showFullScreen()
        else:
            self.showNormal()

    def update_gui(self):     
        if dispatcher.conn is None:
            self.connectionLabel.setText("Not connected")
            self.statisticsLabel.setText("")
            self.old_stats = None
            return

        self.connectionLabel.setText(dispatcher.conn.description)

        # Update RX/TX statistics
        #
        t = datetime.now()

        if self.old_stats is not None:
            dt = (t - self.old_stats.t).total_seconds()
            self.statisticsLabel.setText(
                "RX: %d Byte/s, TX: %d Byte/s" % (
                    (dispatcher.conn.stats.rx_bytes - self.old_stats.rx_bytes) / dt,
                    (dispatcher.conn.stats.tx_bytes - self.old_stats.tx_bytes) / dt
                )
            )

        self.old_stats = copy(dispatcher.conn.stats)
        self.old_stats.t = t

File: QuadControl/MainWindow.cpp
/**
 * Copyright (C)2015 Thomas Kindler <mail_drquad@t-kindler.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "MainWindow.h"

#include "ui_MainWindow.h"

#include <QAction>
#include <QDockWidget>
#include <QSettings>
#include <QCloseEvent>
#include <QMessageBox>
#include <QDebug>
#include "Connection.h"
#include "DockWindows/ConnectionWindow.h"
#include "DockWindows/ConsoleWindow.h"
#include "DockWindows/GLWindow.h"
#include "DockWindows/PlotWindow.h"
#include "DockWindows/UpdateWindow.h"


MainWindow *mainWindow;


MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , connection(this)
    , ui(new Ui::MainWindow)
    , old_stats()
{
    ui->setupUi(this);

    connect(ui->actionExit, &QAction::triggered, this, &MainWindow::close);
    connect(ui->actionAbout, &QAction::triggered, this, &MainWindow::actionAbout_triggered);
    connect(ui->actionAboutQt, &QAction::triggered, this, &MainWindow::actionAboutQt_triggered);
    connect(ui->actionFullscreen, &QAction::triggered, this, &MainWindow::actionFullscreen_triggered);

    ui->statusBar->addPermanentWidget(&statisticsLabel);
    ui->statusBar->addWidget(&connectionLabel);

    connect(&timer, &QTimer::timeout, this, &MainWindow::timer_timeout);
    timer.start(500);
}


MainWindow::~MainWindow()
{
    delete ui;
}


void MainWindow::showEvent(QShowEvent *event)
{
    if (isInitialized || event->spontaneous())
        return;

    connectionWindow = new class ConnectionWindow(this);
    updateWindow     = new class UpdateWindow(this);
    consoleWindow    = new class ConsoleWindow(this);
    glWindow         = new class GLWindow(this);
    plotWindow       = new class PlotWindow(this);

    addDockWindow(Qt::LeftDockWidgetArea, connectionWindow);
    addDockWindow(Qt::LeftDockWidgetArea, updateWindow);

    tabifyDockWidget(
        addDockWindow(Qt::RightDockWidgetArea, glWindow),
        addDockWindow(Qt::RightDockWidgetArea, plotWindow)
    );

    addDockWindow(Qt::BottomDockWidgetArea, consoleWindow);

    QSettings settings;
    restoreGeometry(settings.value("windowGeometry").toByteArray());
    restoreState(settings.value("windowState").toByteArray());

    ui->actionFullscreen->setChecked(windowState() & Qt::WindowFullScreen);

    isInitialized = true;
}


void MainWindow::closeEvent(QCloseEvent *event)
{
    QSettings settings;
    settings.setValue("windowGeometry", saveGeometry());
    settings.setValue("windowState", saveState());
    event->accept();
}


void MainWindow::timer_timeout()
{
    if (connection.isOpen())
        connectionLabel.setText(connection.getUrl().toString());
    else
        connectionLabel.setText("Not connected");

    // Update RX/TX statistics
    //
    auto t = QTime::currentTime();
    auto dt = old_stats_t.msecsTo(t) * 1e-3;

    if (dt > 0) {
        statisticsLabel.setText(
            QString().sprintf("RX: %d Byte/s, TX: %d Byte/s",
                (int)((connection.stats.rx_bytes - old_stats.rx_bytes) / dt),
                (int)((connection.stats.tx_bytes - old_stats.tx_bytes) / dt)
            )
        );
    }

    old_stats = connection.stats;
    old_stats_t = t;
}




QDockWidget *MainWindow::addDockWindow(Qt::DockWidgetArea area, QMainWindow *window)
{
    window->setWindowFlags(Qt::Widget);

    auto dw = new QDockWidget();

    dw->setWidget(window);
    dw->setObjectName(window->windowTitle());
    dw->setWindowTitle(window->windowTitle());
    dw->setAllowedAreas(Qt::AllDockWidgetAreas);

    addDockWidget(area, dw);

    ui->menuView->insertAction(
        ui->menuView->actions().end()[-2],
        dw->toggleViewAction()
    );

    return dw;
}


void MainWindow::actionFullscreen_triggered()
{
    if (ui->actionFullscreen->isChecked())
        this->showFullScreen();
    else
        this->showNormal();
}


void MainWindow::actionAbout_triggered()
{
    QMessageBox::about(this,
        "QuadControl",
        "Copyright (c)2015 Thomas Kindler <a href=\"mailto:mail_drquad@t-kindler.de\">&lt;mail_drquad@t-kindler.de&gt;</a>"
        "<p>"
        "This program is free software: you can redistribute it and/or modify "
        "it under the terms of the GNU General Public License as published by "
        "the Free Software Foundation, either version 3 of the License, or "
        "(at your option) any later version. "
        "<p>"
        "This program is distributed in the hope that it will be useful, "
        "but WITHOUT ANY WARRANTY; without even the implied warranty of "
        "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the "
        "GNU General Public License for more details. "
        "<p>"
        "You should have received a copy of the GNU General Public License "
        "along with this program.  If not, see "
        "<a href=\"http://www.gnu.org/licenses/\">http://www.gnu.org/licenses/</a>."
    );
}


void MainWindow::actionAboutQt_triggered()
{
    QMessageBox::aboutQt(this);
}

